File: src/main/java/io/trydent/olimpo/sys/Id.java
package io.trydent.olimpo.sys;

import java.util.UUID;
import java.util.function.Supplier;

public interface Id extends Supplier<String> {
  static Id id(final String value) {
    return new IdImpl(value);
  }

  static Id uuid() {
    return new Uuid();
  }
}

class IdImpl implements Id {
  private final String value;

  IdImpl(final String value) {
    this.value = value;
  }

  @Override
  public final String get() {
    return this.value;
  }
}

class Uuid implements Id {
  @Override
  public final String get() {
    return UUID.randomUUID().toString();
  }
}

File: src/main/java/io/trydent/olimpo/sys/Id.kt
package io.trydent.olimpo.sys

import java.util.UUID.randomUUID

interface Id : () -> String {
  companion object {
    fun uuid(): Id = RandomUuid()

    fun id(value: String): Id = PlainId(value)
  }
}

internal class RandomUuid : Id {
  override fun invoke(): String = "${randomUUID()}"
}

internal class PlainId(private val value: String) : Id {
  override fun invoke() = value
}


